org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/FinallyFilterTest.java

```diff
@@ -0,0 +1,420 @@
+				"java/lang/Exception");
+		m.visitTryCatchBlock(tryStart, catchEnd, finallyStart, null);
+
+		m.visitLabel(tryStart);
+		m.visitInsn(Opcodes.NOP); // try body
+		m.visitLabel(tryEnd);
+		m.visitJumpInsn(Opcodes.GOTO, finallyEnd);
+
+		m.visitLabel(catchStart);
+		m.visitInsn(Opcodes.POP);
+		m.visitLabel(catchEnd);
+		m.visitInsn(Opcodes.NOP); // finally body
+		shouldMergeLast();
+		m.visitJumpInsn(Opcodes.GOTO, after);
+		shouldIgnoreLast();
+
+		m.visitLabel(finallyStart);
+		m.visitVarInsn(Opcodes.ASTORE, 1);
+		shouldIgnoreLast();
+		m.visitInsn(Opcodes.NOP); // finally body
+		shouldMergeLast();
+		m.visitVarInsn(Opcodes.ALOAD, 1);
+		shouldIgnoreLast();
+		m.visitInsn(Opcodes.ATHROW);
+		shouldIgnoreLast();
+		m.visitLabel(finallyEnd);
+
+		m.visitInsn(Opcodes.NOP); // finally body
+		shouldMergeLast();
+		m.visitLabel(after);
+		m.visitInsn(Opcodes.NOP);
+
+		execute();
+	}
+
+	// === always completes abruptly ===
+
+	@Test
+	public void javac_always_completes_abruptly() {
+		final Label tryStart = new Label();
+		final Label tryEnd = new Label();
+		final Label finallyStart = new Label();
+
+		m.visitTryCatchBlock(tryStart, tryEnd, finallyStart, null);
+
+		m.visitLabel(tryStart);
+		m.visitInsn(Opcodes.NOP); // try body
+		m.visitLabel(tryEnd);
+		m.visitInsn(Opcodes.RETURN); // finally body
+
+		m.visitLabel(finallyStart);
+		m.visitVarInsn(Opcodes.ASTORE, 1);
+		m.visitInsn(Opcodes.RETURN); // finally body
+
+		execute();
+	}
+
+	@Test
+	public void ecj_always_completes_abruptly() {
+		final Label tryStart = new Label();
+		final Label tryEnd = new Label();
+		final Label finallyStart = new Label();
+
+		m.visitTryCatchBlock(tryStart, tryEnd, tryEnd, null);
+
+		m.visitLabel(tryStart);
+		m.visitInsn(Opcodes.NOP); // try body
+		m.visitJumpInsn(Opcodes.GOTO, finallyStart);
+		m.visitLabel(tryEnd);
+
+		m.visitInsn(Opcodes.POP);
+		m.visitLabel(finallyStart);
+		m.visitInsn(Opcodes.RETURN); // finally body
+
+		execute();
+	}
+
+	private final Set<AbstractInsnNode> expectedIgnored = new HashSet<AbstractInsnNode>();
+	private final Set<AbstractInsnNode> actualIgnored = new HashSet<AbstractInsnNode>();
+	private final Set<AbstractInsnNode> expectedMerged = new HashSet<AbstractInsnNode>();
+	private final Set<AbstractInsnNode> actualMerged = new HashSet<AbstractInsnNode>();
+
+	private void shouldMergeLast() {
+		expectedMerged.add(m.instructions.getLast());
+	}
+
+	private void shouldIgnoreLast() {
+		expectedIgnored.add(m.instructions.getLast());
+	}
+
+	private void execute() {
+		filter.filter("", "", m, this);
+		assertEquals("ignored", toIndexes(expectedIgnored),
+				toIndexes(actualIgnored));
+		assertEquals("merged", toIndexes(expectedMerged),
+				toIndexes(actualMerged));
+	}
+
+	private Set<Integer> toIndexes(Set<AbstractInsnNode> set) {
+		final Set<Integer> result = new HashSet<Integer>();
+		for (final AbstractInsnNode i : set) {
+			result.add(m.instructions.indexOf(i));
```

Comment:
@Godin Eclipse shows a boxing warning here.

Reply:
@marchof if replaced by `Integer.valueOf(m.instructions.indexOf(i)))`, then no warning in Eclipse, but warning in IntelliJ with suggestion to remove unnecessary boxing, so added `@SuppressWarnings("boxing")` to this method

Reply:
@Godin Ok. But should give you many warnings at other places then in IntelliJ? ✅

Reply:
@marchof yes, I have quite some warnings in other places in IntelliJ, but we can put this aside